File: Java/ProyectoParejasPOOJava/src/Character.java
import java.util.ArrayList;
import java.util.List;

public abstract class Character {

    public int hp, maxHP, atk, spd, exp, defense, level;
    public String name;
    public boolean isAlive = true;
    public List<CharacterAction> actions = new ArrayList<>();

    public Character(String name) {
        this.name = name;
    }

    public void takeDamage(int damage) {
        int oDefense = 0;
        if (this instanceof Playable p) {
            if (p.isDefending) {
                oDefense  = defense;
                defense  += level - 1;
            }
        }

        int damageTaken = Math.max(0, damage - this.defense);

        if (damageTaken <= 0) {
            BattleUI.displayMiss();
        } else {
            this.hp = Math.max(0, this.hp - damageTaken);
            BattleUI.displayDamage(this, damageTaken);
            if (this instanceof Playable p) {
                if (p.isDefending) {
                    p.isDefending = false;
                    p.defense     = oDefense;
                }
            }
        }

        if (this.hp <= 0) {
            this.isAlive = false;
            if (this instanceof Enemy enemy) {
                if (enemy.onDeath != null)
                    enemy.onDeath.onDeath(enemy.expOnDeath);
            }
        }
    }

    public abstract CharacterAction chooseAction();
}

File: TurnBased-Mess/ProyectoParejasPOO/ProyectoParejasPOO/Character.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ProyectoParejasPOO
{
    public abstract class Character
    {
        public int hp, maxHP, atk, spd, exp, defense, level;
        public string name;
        public bool isAlive = true;
        public List<CharacterAction> actions = new List<CharacterAction>();

        public Character(string name)
        {
            this.name = name;
        }

        public void TakeDamage(int damage)
        {
            int oDefense = 0;
            if (this is Playable) // Si es jugador, revisamos si está defendiendo
            {
                Playable p = (Playable)this;
                if (p.isDefending)
                {
                    //Console.WriteLine("Jugador está defendiendo, aumentando defensa temporalmente."); <-- debug

                    oDefense = defense;
                    defense += (int)(level/2);
                }
            }

            int damageTaken = Math.Max(0, damage - this.defense);
            if (damageTaken <= 0) // Si la defensa anula o supera el daño, no se resta HP, el ataque falla
            {
                BattleUI.DisplayMiss();
            }
            else // De lo contrario, informamos que acierta.
            {
                this.hp = Math.Max(0, this.hp - damageTaken);
                BattleUI.DisplayDamage(this, damageTaken);

                if (this is Playable) // Si es jugador, reseteamos la defensa después de recibir daño
                {
                    Playable p = (Playable)this;
                    if (p.isDefending)
                    {
                        p.isDefending = false; // Reseteamos el estado de defensa después de recibir daño
                        p.defense = oDefense; // Reseteamos la defensa al valor original después de recibir daño

                        //Console.WriteLine("Reiniciado defensa del jugador después de recibir daño."); <-- debug
                    }
                }
            }

            if (this.hp <= 0)
            {
                this.isAlive = false;
                if (this is Enemy)
                {
                    Enemy enemy = (Enemy)this;
                    enemy.OnDeath?.Invoke(enemy.expOnDeath); // Disparamos el evento de muerte para otorgar experiencia al jugador
                }
            }
        }
        public abstract CharacterAction ChooseAction();
    }
}
